### PR TITLE
Add stateful property using parallel workers

### PR DESCRIPTION
**Pull request by pablocostass**

### CODE REVIEW

File: test/segmented_cache_proper_commands.erl [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    IsMember = segmented_cache:is_member(test, Key),
+    case maps:get(Key, State, not_found) of
+        %% If the entry wasn't added, check it's there now
+        not_found ->
+            Entry =:= Value andalso IsMember andalso Res;
+        %% If the entry already existed, check the value was updated
+        ExpectedValue ->
+            Entry =:= Value andalso Entry =/= ExpectedValue andalso IsMember andalso Res
+    end,
+    true;
```
Comment (vkatsuba) [inline test/segmented_cache_proper_commands.erl:103]: Does this functions should always return `true`?

Comment (pablocostass) [reply test/segmented_cache_proper_commands.erl:103]: Ah, good catch! In a draft version of this suite I used asserts to check everything was okay and always returned `true`. This is a leftover from that :)

**Review by NelsonVides — CHANGES_REQUESTED** [review]
Beautiful, I've learned a good lot from this one, never saw a `proper_statem` in action, thank you so much for the brilliant lesson 😄 
I have a few stylistic requests only, then this looks ready to merge 👍🏽

File: test/segmented_cache_proper_commands.erl [hunk 2]
```diff
@@ -0,0 +1,117 @@
+-module(segmented_cache_proper_commands).
+
+%% proper_statem exports
```
[the remark this answers is not in the document]
Comment (pablocostass) [reply test/segmented_cache_proper_commands.erl:3]: I forgot to add it, thankfully PropEr works fine either way :P

File: test/segmented_cache_proper_commands.erl [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+-include_lib("proper/include/proper.hrl").
+-include_lib("stdlib/include/assert.hrl").
+
+initial_state() ->
+    #{}.
+
+command(_State) ->
+    oneof([
+      {call, ?MODULE, is_member, [term()]},
```
[the remark this answers is not in the document]
Comment (pablocostass) [reply test/segmented_cache_proper_commands.erl:25]: Yeah that makes sense, let's avoid running out of atoms :)

File: test/segmented_cache_proper_commands.erl [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    true;
+is_member({next_state, State, _Args, _Res}) ->
+    State;
+is_member({postcondition, State, [Key], Res}) ->
+    ExpectedRes =
+        case maps:get(Key, State, not_found) of
+            not_found -> false;
+            _Value -> true
+        end,
+    ExpectedRes =:= Res;
```
[the remark this answers is not in the document]
Comment (pablocostass) [reply test/segmented_cache_proper_commands.erl:55]: Ah, this is simpler and still readable, I should have thought of it!

File: test/segmented_cache_proper_commands.erl [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+precondition(State, {call, Module, Action, Args}) ->
+    Module:Action({precondition, State, Args}).
+
+next_state(S, Result, {call, Module, Action, Args}) ->
+    Module:Action({next_state, S, Args, Result}).
+
+postcondition(State, {call, Module, Action, Args}, Res) ->
+    ?assert(Module:Action({postcondition, State, Args, Res}),
+            [{command, {Module, Action, Args}}, {result, Res}]),
+    true.
```
[the remark this answers is not in the document]
Comment (pablocostass) [reply test/segmented_cache_proper_commands.erl:41]: Old customs I got from my previous employer. We did that to add even more context to the error log that PropEr would print. But honestly, we also had extra stuff on top of that, so this isn't really necessary.

**Review by NelsonVides — COMMENTED** [review]
Now it looks perfect, but I'm still having trouble to run the tests in my machines. CI passes fine, but locally the one test for the stateful takes 20secs in one group, and it never finishes on the lru group, but without crashing the VM. Any idea why could that be? Trying to debug my beam 🤔

**Comment by NelsonVides** [comment]
Actually, tests do finish, what fails is to finish producing the ct html files. They start, until `segmented_cache_SUITE | lru | stateful_property`, then the html files are left behind 🤔

**Comment by pablocostass** [comment]
Wait, what? That CT issue is something I never heard of :scream:

As for the previous comment, if you find yourself in that situation again try changing the number of workers to `0`, `1` and `2` to see where the issue lays (is the property itself the problem, is the code that handles the worker the problem, or is it a concurrent issue?).

**Comment by NelsonVides** [comment]
> Wait, what? That CT issue is something I never heard of 😱
> 
> As for the previous comment, if you find yourself in that situation again try changing the number of workers to `0`, `1` and `2` to see where the issue lays (is the property itself the problem, is the code that handles the worker the problem, or is it a concurrent issue?).

It works fine if I set the number of tests to 10k instead of the 100k default, or if I set a low number of workers, of if I run this test alone without all others. It seems like some ct worker responsible from building the html might be just dying from the load, funny. Need to debug a bit 😄

**Comment by pablocostass** [comment]
Keep me updated, now I am intrigued on what's happening :joy:

**Review by NelsonVides — APPROVED** [review]
Absolutely gorgeous PR 😄 
I can get it merged, it passes CI here every single time, so maybe it's about my systems. Will figure out, this one is great as it is 😄

**Pull request merged**